File: LeetcodeEasyProblems/_141. Linked List Cycle/Solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        # First define the fast and slow pointers
        # and put them at the start of the list
        slow, fast = head, head
        # The speed of the slow pointer will be 1 nodes
        # per iteration, while the fast pointer will be 2

        # If they don't have a cycle, the fast pointer
        # will just reach the end of the linked list first

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        
        return False


        

File: LeetcodeEasyProblems/_141. Linked List Cycle/Solution.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        // We tackle the case when the head is null
        if (head == null){
            return false;
        }
        // We create the slow and fast pointers
        ListNode slow = head;
        ListNode fast = head;
        // Use the while loop to check
        // if the next Node and the second node from the 
        // current fast pointers are not null (the end of the list)
        while(fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next.next;
            if (slow.equals(fast)){
                return true;
            }

        }
        // If the fast pointer simply reach the end
        // of the linked list first, then
        //there's no cycle
        return false;

        
    }
}